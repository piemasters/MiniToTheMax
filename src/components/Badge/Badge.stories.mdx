import {
  Story,
  Preview,
  Props,
  Source,
  Description,
  Meta,
} from '@storybook/addon-docs/blocks';
import { Badge } from './Badge';

<Meta
  title="Components/Badge"
  component={Badge}
  parameters={{ jest: ['badge'] }}
  argTypes={{
    type: {
      name: 'type',
      description: 'The badge type to display',
      defaultValue: 'showcase',
      control: { type: 'text' },
    },
  }}
/>

# Badge

With `MDX` we can define a story for `Badge` right in the middle of our
markdown documentation.

<Preview>
  <Story
    name="default"
    args={{
      type: 'showcase',
    }}
  >
    {(args) => <Badge {...args} />}
  </Story>
</Preview>

# Props

<Props story="default" />
