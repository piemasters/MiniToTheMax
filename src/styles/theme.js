export default {
  colors: {
    text: '#000',
    muted: '#999',
  },
};
