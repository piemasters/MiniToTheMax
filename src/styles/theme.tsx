export interface Theme {
  colors: {
    text: string;
    muted: string;
  };
}

const appTheme: Theme = {
  colors: {
    text: '#000',
    muted: '#999',
  },
};

export default appTheme;
