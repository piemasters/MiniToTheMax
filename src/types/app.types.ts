import type { IGatsbyImageData } from 'gatsby-plugin-image';

import { AllMdx, MdxGroupConnection, MdxNode, Site } from './base.types';
import { TechnicalPaintImages } from '../data/paints/technical';

export interface Posts {
  posts: AllMdx;
}

export interface PostSummary {
  slug: string;
  title: string;
  date: string;
  img: IGatsbyImageData;
  excerpt: string;
}

export interface MdxPost {
  site: Site;
  post: MdxNode;
}

export interface Categories {
  categories: AllMdx;
}

export interface Showcase {
  boardGames: IGatsbyImageData;
  gloomspiteGitz: IGatsbyImageData;
  spaceWolves: IGatsbyImageData;
  greyKnights: IGatsbyImageData;
  orks: IGatsbyImageData;
  scenery: IGatsbyImageData;
}

export interface Tutorials {
  fortyThousand: IGatsbyImageData;
  bases: IGatsbyImageData;
  scenery: IGatsbyImageData;
  tools: IGatsbyImageData;
}

export interface Reviews {
  boardGames: IGatsbyImageData;
  books: IGatsbyImageData;
  tools: IGatsbyImageData;
}

export interface BattleReports {
  fortyThousand: IGatsbyImageData;
}

export type Category = MdxGroupConnection;

export interface Tags {
  tags: AllMdx;
}

export type Tag = MdxGroupConnection;

export interface TagLink {
  slug: string;
  title: string;
}

export interface PostLink extends TagLink {
  img: IGatsbyImageData;
}

export interface DisqusConfig {
  identifier: string;
  title: string;
  url: string;
}

export enum BadgeNames {
  tutorials = 'tutorials',
  showcase = 'showcase',
  build = 'build',
  battleReports = 'battleReports',
  reviews = 'reviews',
}

export interface PaintGradient {
  offset: number;
  color: string;
}

export interface PaintDetails {
  [key: string]: string | boolean | PaintGradient[] | JSX.Element | undefined;
  name: string;
  type: string;
  color: string;
  hex: string;
  gradient?: PaintGradient[];
  stroke?: string;
  gloss?: boolean;
  img?: TechnicalPaintImages;
  availability: string;
}
export interface PaintDetailsComponent extends Omit<PaintDetails, 'img'> {
  img?: JSX.Element;
}

export interface NavLink {
  name: string;
  url: string;
}

export interface BacklogEntry {
  name: string;
  link?: string;
}

export interface PaintFilters {
  [key: string]: boolean;
}

export interface AllPaintFilters {
  [key: string]: PaintFilters;
}
