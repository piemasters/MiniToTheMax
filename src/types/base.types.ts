import type { IGatsbyImageData } from 'gatsby-plugin-image';

export type SiteSiteMetadata = {
  title: string;
  author: string;
  description: string;
  siteUrl: string;
  image: string;
};

export interface Site {
  id: string;
  parent: BaseNode;
  children: [BaseNode];
  internal: Internal;
  siteMetadata: SiteSiteMetadata;
  port: number;
  host: string;
  polyfill: boolean;
  pathPrefix: string;
  buildTime: string;
}

export interface MdxNode {
  rawBody: string;
  fileAbsolutePath: string;
  frontmatter: MdxFrontmatter;
  body: string;
  excerpt: string;
  headings: [MdxHeadingMdx];
  html: string;
  mdxAST: JSON;
  tableOfContents: JSON;
  wordCount: MdxWordCount;
  fields: MdxFields;
  id: string;
  parent: BaseNode;
  children: [BaseNode];
  internal: Internal;
}

export type MdxHeadingMdx = {
  value: string;
  depth: number;
};

export type MdxWordCount = {
  paragraphs: number;
  sentences: number;
  words: number;
};

export type MdxFields = {
  slug: string;
};

export type Internal = {
  content: string;
  contentDigest: string;
  description: string;
  fieldOwners: [string];
  ignoreType: boolean;
  mediaType: string;
  owner: string;
  type: string;
};

export interface BaseNode {
  id: string;
  parent: BaseNode;
  children: [BaseNode];
  internal: Internal;
}

export type PageInfo = {
  currentPage: number;
  hasPreviousPage: boolean;
  hasNextPage: boolean;
  itemCount: number;
  pageCount: number;
  perPage: number;
};

export type MdxGroupConnection = {
  totalCount: number;
  edges: [MdxEdge];
  nodes: [MdxNode];
  pageInfo: PageInfo;
  field: string;
  fieldValue: string;
};

export interface MdxEdge {
  next: MdxNode;
  node: MdxNode;
  previous: MdxNode;
}

export interface AllMdx {
  totalCount: number;
  edges: [MdxEdge];
  nodes: [MdxNode];
  pageInfo: PageInfo;
  distinct: [string];
  group: [MdxGroupConnection];
}

export interface ImageProp {
  full: IGatsbyImageData;
  thumb: IGatsbyImageData;
  thumbAlt?: string;
  title?: string;
  caption?: string;
}

export interface MdxFrontmatterGalleryImage {
  publicURL: string;
  childImageSharp: ImageProp;
}

export interface MdxFrontmatter {
  title: string;
  date: string;
  categories: [string];
  tags: [string];
  featuredImage: {
    publicURL?: string;
    childImageSharp: {
      gatsbyImageData: IGatsbyImageData;
    };
  };
  gallery: [MdxFrontmatterGalleryImage];
  published: boolean;
}
