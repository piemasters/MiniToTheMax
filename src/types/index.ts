export * from './app.types';
export * from './base.types';
