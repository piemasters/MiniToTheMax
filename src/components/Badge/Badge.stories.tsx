import type { Meta, StoryObj } from '@storybook/react';

import { Badge } from './Badge';
import { BadgeNames } from '../../types';

const meta: Meta<typeof Badge> = {
  component: Badge,
  argTypes: {
    type: {
      name: 'type',
      description: 'The badge type to display',
      defaultValue: BadgeNames.showcase,
      control: { type: 'radio' },
    },
  },
};

export default meta;
type Story = StoryObj<typeof Badge>;

export const Primary: Story = {
  args: {
    type: BadgeNames.showcase,
  },
};
