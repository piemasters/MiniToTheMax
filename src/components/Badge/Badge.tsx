import type { FC } from 'react';
import { css } from '@emotion/react';

import { BadgeNames } from '../../types';
import { allBadges } from '../../data';

export interface BadgeProps {
  type: BadgeNames;
}

export const Badge: FC<BadgeProps> = ({ type }) => {
  const badgeStyles = css`
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    width: 9rem; // 7rem for 5
  `;

  const badgeImageStyles = css`
    margin: 0 auto 0.4rem;
    width: 75%;
    opacity: 0.9;
    &:hover {
      opacity: 1;
    }
  `;

  const badgeTextStyles = css`
    text-align: center;
    text-transform: capitalize;
  `;

  return (
    <div css={badgeStyles} data-testid="badge">
      <img src={allBadges[type]} alt="" css={badgeImageStyles} />
      <p css={badgeTextStyles}>{type.replace(/([A-Z])/g, ' $1').trim()}</p>
    </div>
  );
};

export default Badge;
