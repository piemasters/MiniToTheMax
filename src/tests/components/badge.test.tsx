import React from 'react';
import Badge from '../../components/badge';
import { getByTestId, render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import theme from '../../styles/theme';
import { BadgeNames } from '../../types/app.types';

describe('Badge', () => {
  test('renders correctly', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Badge type={BadgeNames.showcase} />
      </ThemeProvider>
    );

    expect(getByTestId(container, 'badge')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
