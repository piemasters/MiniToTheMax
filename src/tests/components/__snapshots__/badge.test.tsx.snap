// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders correctly 1`] = `
<div>
  <div
    css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
    data-testid="badge"
  >
    <img
      alt=""
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      src="test-file-stub"
    />
    <p
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
    >
      showcase
    </p>
  </div>
</div>
`;
